avocado/plugins/beaker_result.py

```diff
@@ -0,0 +1,137 @@
+            return
+        baseurl = baseurl.rstrip('/')
+        self.beaker_url = baseurl + '/recipes/' + recipeid + '/tasks/' + taskid
+        LOG_UI.info("beaker: using API at %s (R:%s T:%s)", baseurl, recipeid, taskid)
+
+    def send_request(self, req):
+        LOG_UI.info('beaker: %s %s ...', req.method, req.full_url)
+        try:
+            res = urllib.request.urlopen(req)  # nosec
+            return res
+        except urllib.error.URLError as err:
+            LOG_UI.info('beaker: Failed: %s', err)
+            return None
+        except Exception as err:
+            # should not happen
+            LOG_UI.info('beaker: Oops: %s', err)
+            return None
+
+    def post_result(self, state):
+        reqdict = {
+            'path': "%s/%s" % (self.job_id, state.get('name')),
+        }
+
+        result = state.get('status').lower().capitalize()
+        if result == 'Cancel':
+            result = 'Skip'
+        if result not in ['Pass', 'Fail', 'Skip']:
+            result = 'Fail'
+        reqdict['result'] = result
+
+        reason = state.get('fail_reason')
+        if reason is not None:
+            reqdict['message'] = reason
+
+        secs = state.get('time_elapsed')
+        if secs is not None:
+            reqdict['score'] = str(int(secs))
+
+        reqdata = urllib.parse.urlencode(reqdict).encode('utf-8')
+        url = self.beaker_url + '/results/'
+        req = urllib.request.Request(url, method='POST', data=reqdata)
+        res = self.send_request(req)
+        if res is None:
+            return None
+        return res.getheader('Location')
+
+    def put_data(self, location, name, content):
+        url = location + '/logs/' + name
+        req = urllib.request.Request(url, method='PUT', data=content)
+        self.send_request(req)
+
+    def put_file(self, location, name, filename):
+        file = open(filename)
+        content = file.read().encode('utf-8')
+        file.close()
+        self.put_data(location, name, content)
+
+    def put_file_list(self, location, prefix, filelist):
+        for file in filelist:
+            if os.path.isfile(file) and os.path.getsize(file) > 0:
+                name = prefix + os.path.basename(file)
+                self.put_file(location, name, file)
+
+    def pre_tests(self, job):
+        self.job_id = job.unique_id[:6]
+        return
+
+    def start_test(self, result, state):
+        return
+
+    def test_progress(self, progress=False):
+        return
+
+    def end_test(self, result, state):
+        if self.beaker_url is None:
+            return
+
+        location = self.post_result(state)
+        if location is None:
+            return
+
+        logfile = state.get('logfile')
+        self.put_file(location, 'logfile', logfile)
+
+        ppstate = pprint.pformat(state).encode('utf8')
+        self.put_data(location, 'state', ppstate)
+
+        pattern = os.path.join(state.get('logdir'), '*')
```

Comment:
Interesting... I wonder if we should keep this "last state" file as a standard part of the results.  We have similar information in the `result.json` file, but that one is job-wide.

Reply:
Did this because I can't store some fields in beaker, state->fail_reason for example (I store that as 'message' but it seems to not show up in the beaker web ui).
It's also useful for trouble-shooting (state->traceback).
But, yes, I can see that being useful as general service, i.e. store this as state.txt (or state.json) in logdir.